Generic promise batching implementation

poltak opened the pull request:
Mostly relates to what is being discussed in #28 regarding batch running of a fetch&analyse function. Work on that function is being done elsewhere, so instead this is a generic batching logic that will be able to be used with fetch&analyse function (and any other async function).

_More information (from original discussion on [WorldBrain PR #15](https://github.com/WorldBrain/WebMemex/pull/15) - __when batching logic was coupled to fetch&analyse__):_

Spent the last day getting familiar with RxJS usage. Used elsewhere in this project and heard about it enough lately to prompt me to play with it. Interesting stuff. Anyway, I got the idea of maybe trying to write the batch fetch&analyse logic using an Observable. 

This may be completely the wrong direction, as my understanding may be very flawed at this point, however I managed to get something working that was surprisingly simple (code-wise; still took ages to get working) and works (at a basic level so far).

### Pre-knowledge

The previous work on this PR essentially resolved in: `store-page` can now work with and without a browser tab. Meaning you can just pass a URL and have all the dedup, data fetching + analysis logic run on it. The `store-page` interface takes `url` and returns a `Promise`. This `Promise` eventually creates a page stub and resolves to an object containing:

- `page` (contains the `_id` of the inserted page stub + `url`)
- `finalPagePromise` (another promise that does all the analysis and updates the page stub, if needed)

The first promised returned by the `store-page` interface usually resolves pretty fast, as it's just doing a DB op. The `finalPagePromise` usually takes a bit longer, as it requires doing an XHR and some other stuff. 

In my mind, the batch/whatever should be able to take a bunch of URLs and then batch up these `finalPagePromise`s, waiting on _n_ at any time. It should also be able to stop anytime and pick-up again from where it left off. 

### Idea

The native ES6 Promise (and asyncawait) API doesn't seem powerful enough on its own to do this without a lot of extra code. However I found [`mergeMap`](https://www.learnrxjs.io/operators/transformation/mergemap.html) in RxJS that seems to be able to do this coupled with [`defer`](https://github.com/Reactive-Extensions/RxJS/blob/master/doc/api/core/operators/defer.md) to stop all the promises from running at the same time (limits to the concurrency level specified). 

On a stream of URLs, `mergeMap` takes each one and defers the `finalPagePromise` returned from calling that URL on the `store-page` interface. Concurrency level will allow _n_ promises to run at anytime while the others are still deferred until the currently waiting promises get resolved.

An internal state is stored, which currently URLs get written to as they get pushed off the `mergeMap` stream, meaning their `finalPagePromise` has resolved. This allows restarting of the batch as long as the state is stored somewhere (either internally in the batch function, or externally if the function needs to be free'd).

### Confused part

Because the `store-page` interface returns a promise which resolves to a bit of data + another promise, I am currently doing some weird stuff to get that __inner__ promise to make sure the `Observable.defer` gets it (instead of outer promise). It means I can't emit other information (like the page ID or URL) to the streams (or at least I don't understand how). example:

```js
observable.mergeMap(
    input => Rx.Observable.defer(
        () => returnsNestedPromise(input).then(
            res => res.innerPromise)))
```

### Example

Here's an example CodePen I was playing with to understand everything. It runs the same batching logic, but with a mocked fetch&analyse function, that also logs everything to console to see how it works: https://codepen.io/poltak/pen/XRRdeR

poltak committed:
Write basic concurrent batch logic for fetch&analyse

- plug into the existing reidentifyOrStorePage (now that it doesn't need tab)
- use RxJS observable's mergeMap to defer all analysis promises and have them run concurrently

poltak committed:
Add pause/resume to URL batch interface

- also adds internal state to track what URLs already processed
- on resume, does an diff on processed data state and init data state to ignore already-processed URLs
- add getRemaining() to interface to access progress

poltak committed:
Make batching logic work for any async function

- now exports a HOF that binds the async function to the batch initialiser; user can then initialise it with some input then as before

poltak committed:
Rename batch-store-ulrs to promise-batch

- moved from page-storage to util as it's now generic and not coupled to page storage

poltak committed:
Add terminate to batch handler interface

- terminate needs to clear state, which unsubbing won't
- rename getRemaining to getProgress and just return the processed state for the caller to act on

poltak committed:
Clean up pauser logic

- no need two subjects; just have single `pauser` created, which all the interface methods act on
- need to pass in the observable as a callback because the set difference (progress resuming) needs to happen each time

Treora commented:
Nice challenge, to make a pausable promise executor using Rx. It is ideal if the complexity is contained in the file while the outside interface is simple. I wonder how exactly this would be used in the application. 

I think you took out the struggle with the promise-in-a-promise construct? This whole construct (that I made for deduplication) is not so pretty and could be changed; actually an Observable may be a cleaner way for such a twice-resolving promise; or the `finalPagePromise` may not be needed at all (I suppose we could listen to database changes instead).

I can't yet fully wrap my head around it though.

poltak committed:
Clean up resume state checking logic in batcher

- was manually getting the diff of two sets
- this could just be done in the observable stream with a .filter call
- btw, this now allows the input as any Iterable, array-like, or Observer, instead of just a Set, which is powerful
- at the moment the pausing logic will only work with an Iterable/array-like containing primitive vals due to the filter using Set.has() method

poltak committed:
Add support to inner-promises to promise-batcher

- this changes the stream ops a bit
- now the callback gets mapped to the stream that gets fed into the mergeMap; the mergeMap expects the callback to return a promise, hence it will operate on a stream of promises

poltak committed:
Make sure to propagate input to end of batch stream

- now explicitly passes the input at each stage to the next to make sure we can grab it at the end
- this is because the input is stored at the end to track which inputs have been processed (filtered at earlier stage)
- little doc note about this as it makes things messier

poltak commented:
> I think you took out the struggle with the promise-in-a-promise construct? This whole construct (that I made for deduplication) is not so pretty and could be changed

I've left it as is right now, but I'm thinking instead of returning the `finalPagePromise`, return a function that returns `finalPagePromise`. The reason is that Promises will start to process as soon as they're created, and enclosing the creation in a function will essentially defer the Promise until it's called. So I can call a `Rx.Observable.defer` on that function, but if I try to defer on the Promise that's already created, it may be already resolved by the time it gets to that part of the stream (making the idea of "promise-batching" useless, as there's no control over which promises are allowed to resolve at any time). Not sure if that makes sense...

> actually an Observable may be a cleaner way for such a twice-resolving promise

Added optional support for these cases in 0dc70b7. Still has main `callback` arg which sets up the async function to batch, but if that returns another async func/Promise, it'll be handled. (In Rx, need to use `.flatMap` operator rather than `.map` for these cases).

poltak committed:
Fix promise-batcher terminate interface method

- complete signal wasn't completely terminating it
- pushing false to the subject first will make sure the subject isn't using input observable
- put line-by-line doc

poltak committed:
Overhal promise-batcher's state management

- previously was based on an ES6 Set
- problem was that Sets only work properly with primitive vals, hence for any input containing objects, pause + resume wouldn't work properly
- now based on an Array + logic to enforce Set constraints and compare any types to allow batcher to accept generic input types

poltak committed:
Allow full observer to be passed into batch subscribe

- Rx5 needs three args to subscribe to react to the different events

poltak committed:
Overhaul promise-batcher pausing implementation

- previous pausing implementation worked alright, but it was the wrong way to go about it
- it meant subscriber couldn't act on completed events
- this way essentially stops the observable all-together, by unsubscribing, and when user resubs the state is used to filter out changes
- subscriber is actually subbing to the subject of the original observable, but now that subject is subscribed to the observable and gets the events
- also fixed problem in page-storage batch

poltak committed:
Change batching logic (yet again) to address design flaw

- the whole innerPromise support seemed to be built on my misunderstanding
- this would originally run every outer promise (no defer + concurrency), hence make batcher space complexity dependent on input size
- the point of the batcher is to make a constant space overhead by running a constant number of input at once, so that feature was silly
- why not just support async fns/promises and let the caller organise their logic in a single fn which can then be batched
- Rx stuff now hidden away via interface, as well
- original observable stream now more closely resembles original batcher I wrote

poltak committed:
Update batcher to allow observer fns to be passed at init

- start() function shouldn't really have to do that
- makes anyone be able to init a batch and observe it some how while passing interface to someone else

poltak committed:
Improve error handling on promise batcher

- now the stream won't stop if an error is caught
- batcher will run observer's onError callback (caller specified) with the err, plus mark it off in state as error
- put the defer observable in a separate var for clarity
- removed the binding interface of the module; caller can specify all args at once
